File: backend/src/HIS.Core/Entities/IvfLab.cs
namespace HIS.Core.Entities;

/// <summary>
/// Cặp vợ chồng IVF (In Vitro Fertilization)
/// </summary>
public class IvfPatientCouple : BaseEntity
{
    public Guid WifePatientId { get; set; }
    public Guid HusbandPatientId { get; set; }
    public int InfertilityDurationMonths { get; set; }
    public string? InfertilityCause { get; set; }
    public DateTime? MarriageDate { get; set; }
    public string? Notes { get; set; }

    // Navigation
    public virtual Patient? WifePatient { get; set; }
    public virtual Patient? HusbandPatient { get; set; }
    public virtual ICollection<IvfCycle> Cycles { get; set; } = new List<IvfCycle>();
}

/// <summary>
/// Chu kỳ IVF
/// Status: 1=Active, 2=OvumPickup, 3=Fertilization, 4=Transfer, 5=Frozen, 6=Completed, 7=Cancelled
/// </summary>
public class IvfCycle : BaseEntity
{
    public Guid CoupleId { get; set; }
    public int CycleNumber { get; set; }
    public DateTime StartDate { get; set; }
    public int Status { get; set; } = 1;
    public string? Protocol { get; set; }
    public Guid? DoctorId { get; set; }
    public string? Notes { get; set; }

    // Navigation
    public virtual IvfPatientCouple? Couple { get; set; }
    public virtual User? Doctor { get; set; }
    public virtual IvfOvumPickup? OvumPickup { get; set; }
    public virtual ICollection<IvfEmbryo> Embryos { get; set; } = new List<IvfEmbryo>();
    public virtual ICollection<IvfEmbryoTransfer> Transfers { get; set; } = new List<IvfEmbryoTransfer>();
    public virtual ICollection<IvfBiopsy> Biopsies { get; set; } = new List<IvfBiopsy>();
}

/// <summary>
/// Chọc trứng (Ovum Pickup / OPU)
/// </summary>
public class IvfOvumPickup : BaseEntity
{
    public Guid CycleId { get; set; }
    public DateTime PickupDate { get; set; }
    public int TotalOvums { get; set; }
    public int MatureOvums { get; set; }
    public int ImmatureOvums { get; set; }
    public int DegeneratedOvums { get; set; }
    public Guid? PerformedById { get; set; }
    public string? Notes { get; set; }

    // Navigation
    public virtual IvfCycle? Cycle { get; set; }
    public virtual User? PerformedBy { get; set; }
}

/// <summary>
/// Phôi (Embryo)
/// Status: 1=Culture, 2=Fresh_Transfer, 3=Frozen, 4=Thawed, 5=Transferred, 6=Discarded
/// </summary>
public class IvfEmbryo : BaseEntity
{
    public Guid CycleId { get; set; }
    public string EmbryoCode { get; set; } = string.Empty;
    public string? Day2Grade { get; set; }
    public string? Day3Grade { get; set; }
    public string? Day5Grade { get; set; }
    public string? Day6Grade { get; set; }
    public string? Day7Grade { get; set; }
    public int Status { get; set; } = 1;
    public DateTime? FreezeDate { get; set; }
    public DateTime? ThawDate { get; set; }
    public string? StrawCode { get; set; }
    public string? StrawColor { get; set; }
    public string? BoxCode { get; set; }
    public string? TankCode { get; set; }
    public string? RackPosition { get; set; }
    public string? Notes { get; set; }
    public string? ImageUrl { get; set; }

    // Navigation
    public virtual IvfCycle? Cycle { get; set; }
}

/// <summary>
/// Chuyển phôi (Embryo Transfer)
/// TransferType: 1=Fresh, 2=Frozen
/// ResultStatus: 0=Pending, 1=Positive, 2=Negative
/// </summary>
public class IvfEmbryoTransfer : BaseEntity
{
    public Guid CycleId { get; set; }
    public DateTime TransferDate { get; set; }
    public int TransferType { get; set; } = 1;
    public int EmbryoCount { get; set; }
    public Guid? DoctorId { get; set; }
    public Guid? EmbryologistId { get; set; }
    public string? Notes { get; set; }
    public int ResultStatus { get; set; }

    // Navigation
    public virtual IvfCycle? Cycle { get; set; }
    public virtual User? Doctor { get; set; }
    public virtual User? Embryologist { get; set; }
}

/// <summary>
/// Ngân hàng tinh trùng (Sperm Bank)
/// Status: 1=Stored, 2=Used, 3=Disposed
/// </summary>
public class IvfSpermBank : BaseEntity
{
    public Guid PatientId { get; set; }
    public string SampleCode { get; set; } = string.Empty;
    public DateTime CollectionDate { get; set; }
    public decimal? Volume { get; set; }
    public decimal? Concentration { get; set; }
    public decimal? Motility { get; set; }
    public decimal? Morphology { get; set; }
    public int StrawCount { get; set; }
    public string? TankCode { get; set; }
    public string? RackPosition { get; set; }
    public string? BoxCode { get; set; }
    public int Status { get; set; } = 1;
    public DateTime? ExpiryDate { get; set; }
    public decimal StorageFee { get; set; }
    public string? Notes { get; set; }

    // Navigation
    public virtual Patient? Patient { get; set; }
}

/// <summary>
/// Sinh thiết phôi (Embryo Biopsy / PGT)
/// </summary>
public class IvfBiopsy : BaseEntity
{
    public Guid CycleId { get; set; }
    public Guid? PatientId { get; set; }
    public string? BiopsyLab { get; set; }
    public DateTime? SentDate { get; set; }
    public DateTime? ResultDate { get; set; }
    public string? Result { get; set; }
    public string? Notes { get; set; }

    // Navigation
    public virtual IvfCycle? Cycle { get; set; }
    public virtual Patient? Patient { get; set; }
}

File: frontend/src/api/ivfLab.ts
import { apiClient } from './client';

// ---- Types ----

export interface IvfCouple {
  id: string;
  wifePatientId: string;
  wifeName?: string;
  wifeCode?: string;
  wifeDob?: string;
  husbandPatientId: string;
  husbandName?: string;
  husbandCode?: string;
  husbandDob?: string;
  infertilityDurationMonths: number;
  infertilityCause?: string;
  marriageDate?: string;
  notes?: string;
  cycleCount: number;
}

export interface IvfCoupleDetail extends IvfCouple {
  cycles: IvfCycle[];
}

export interface IvfCycle {
  id: string;
  coupleId: string;
  cycleNumber: number;
  startDate?: string;
  status: number;
  statusName?: string;
  protocol?: string;
  doctorId?: string;
  doctorName?: string;
  notes?: string;
  embryoCount: number;
  transferCount: number;
}

export interface IvfCycleDetail extends IvfCycle {
  ovumPickup?: IvfOvumPickup;
  embryos: IvfEmbryo[];
  transfers: IvfTransfer[];
  biopsies: IvfBiopsy[];
}

export interface IvfOvumPickup {
  id: string;
  cycleId: string;
  pickupDate?: string;
  totalOvums: number;
  matureOvums: number;
  immatureOvums: number;
  degeneratedOvums: number;
  performedById?: string;
  performedByName?: string;
  notes?: string;
}

export interface IvfEmbryo {
  id: string;
  cycleId: string;
  embryoCode: string;
  day2Grade?: string;
  day3Grade?: string;
  day5Grade?: string;
  day6Grade?: string;
  day7Grade?: string;
  status: number;
  statusName?: string;
  freezeDate?: string;
  thawDate?: string;
  strawCode?: string;
  strawColor?: string;
  boxCode?: string;
  tankCode?: string;
  rackPosition?: string;
  notes?: string;
  imageUrl?: string;
}

export interface IvfTransfer {
  id: string;
  cycleId: string;
  transferDate?: string;
  transferType: number;
  transferTypeName?: string;
  embryoCount: number;
  doctorId?: string;
  doctorName?: string;
  embryologistId?: string;
  embryologistName?: string;
  notes?: string;
  resultStatus: number;
  resultStatusName?: string;
}

export interface IvfSpermSample {
  id: string;
  patientId: string;
  patientName?: string;
  patientCode?: string;
  sampleCode: string;
  collectionDate?: string;
  volume?: number;
  concentration?: number;
  motility?: number;
  morphology?: number;
  strawCount: number;
  tankCode?: string;
  rackPosition?: string;
  boxCode?: string;
  status: number;
  statusName?: string;
  expiryDate?: string;
  storageFee: number;
  notes?: string;
}

export interface IvfBiopsy {
  id: string;
  cycleId: string;
  patientId?: string;
  patientName?: string;
  biopsyLab?: string;
  sentDate?: string;
  resultDate?: string;
  result?: string;
  notes?: string;
}

export interface IvfDashboard {
  activeCycles: number;
  frozenEmbryos: number;
  spermSamples: number;
  transfersThisMonth: number;
  successRate: number;
  totalCouples: number;
  completedCycles: number;
}

export interface IvfDailyReport {
  date?: string;
  items: { activityType: string; count: number; details?: string }[];
}

// ---- API Functions ----

export const getCouples = async (params?: { keyword?: string; pageIndex?: number; pageSize?: number }) => {
  try {
    const response = await apiClient.get<IvfCouple[]>('/ivf-lab/couples', { params });
    return response.data || [];
  } catch {
    console.warn('Failed to fetch IVF couples');
    return [];
  }
};

export const getCoupleById = async (id: string) => {
  try {
    const response = await apiClient.get<IvfCoupleDetail>(`/ivf-lab/couples/${id}`);
    return response.data;
  } catch {
    console.warn('Failed to fetch IVF couple detail');
    return null;
  }
};

export const saveCouple = async (data: Partial<IvfCouple>) => {
  const response = await apiClient.post<IvfCouple>('/ivf-lab/couples', data);
  return response.data;
};

export const getCycles = async (coupleId: string) => {
  try {
    const response = await apiClient.get<IvfCycle[]>('/ivf-lab/cycles', { params: { coupleId } });
    return response.data || [];
  } catch {
    console.warn('Failed to fetch IVF cycles');
    return [];
  }
};

export const getCycleById = async (id: string) => {
  try {
    const response = await apiClient.get<IvfCycleDetail>(`/ivf-lab/cycles/${id}`);
    return response.data;
  } catch {
    console.warn('Failed to fetch IVF cycle detail');
    return null;
  }
};

export const saveCycle = async (data: Partial<IvfCycle>) => {
  const response = await apiClient.post<IvfCycle>('/ivf-lab/cycles', data);
  return response.data;
};

export const updateCycleStatus = async (id: string, status: number) => {
  await apiClient.put(`/ivf-lab/cycles/${id}/status`, { status });
};

export const saveOvumPickup = async (data: Partial<IvfOvumPickup>) => {
  const response = await apiClient.post<IvfOvumPickup>('/ivf-lab/ovum-pickup', data);
  return response.data;
};

export const getOvumPickup = async (cycleId: string) => {
  try {
    const response = await apiClient.get<IvfOvumPickup>('/ivf-lab/ovum-pickup', { params: { cycleId } });
    return response.data;
  } catch {
    console.warn('Failed to fetch ovum pickup');
    return null;
  }
};

export const getEmbryos = async (cycleId: string) => {
  try {
    const response = await apiClient.get<IvfEmbryo[]>('/ivf-lab/embryos', { params: { cycleId } });
    return response.data || [];
  } catch {
    console.warn('Failed to fetch embryos');
    return [];
  }
};

export const saveEmbryo = async (data: Partial<IvfEmbryo>) => {
  const response = await apiClient.post<IvfEmbryo>('/ivf-lab/embryos', data);
  return response.data;
};

export const updateEmbryoStatus = async (id: string, status: number) => {
  await apiClient.put(`/ivf-lab/embryos/${id}/status`, { status });
};

export const freezeEmbryo = async (id: string, data: { freezeDate?: string; strawCode?: string; strawColor?: string; boxCode?: string; tankCode?: string; rackPosition?: string }) => {
  await apiClient.put(`/ivf-lab/embryos/${id}/freeze`, data);
};

export const thawEmbryo = async (id: string, data: { thawDate?: string }) => {
  await apiClient.put(`/ivf-lab/embryos/${id}/thaw`, data);
};

export const getTransfers = async (cycleId: string) => {
  try {
    const response = await apiClient.get<IvfTransfer[]>('/ivf-lab/transfers', { params: { cycleId } });
    return response.data || [];
  } catch {
    console.warn('Failed to fetch transfers');
    return [];
  }
};

export const saveTransfer = async (data: Partial<IvfTransfer>) => {
  const response = await apiClient.post<IvfTransfer>('/ivf-lab/transfers', data);
  return response.data;
};

export const updateTransferResult = async (id: string, resultStatus: number) => {
  await apiClient.put(`/ivf-lab/transfers/${id}/result`, { resultStatus });
};

export const getSpermSamples = async (params?: { keyword?: string; status?: number; pageIndex?: number; pageSize?: number }) => {
  try {
    const response = await apiClient.get<IvfSpermSample[]>('/ivf-lab/sperm-bank', { params });
    return response.data || [];
  } catch {
    console.warn('Failed to fetch sperm samples');
    return [];
  }
};

export const saveSpermSample = async (data: Partial<IvfSpermSample>) => {
  const response = await apiClient.post<IvfSpermSample>('/ivf-lab/sperm-bank', data);
  return response.data;
};

export const updateSpermStatus = async (id: string, status: number) => {
  await apiClient.put(`/ivf-lab/sperm-bank/${id}/status`, { status });
};

export const getExpiringStorage = async (daysAhead = 30) => {
  try {
    const response = await apiClient.get<IvfSpermSample[]>('/ivf-lab/sperm-bank/expiring', { params: { daysAhead } });
    return response.data || [];
  } catch {
    console.warn('Failed to fetch expiring storage');
    return [];
  }
};

export const getBiopsies = async (params?: { cycleId?: string; patientId?: string }) => {
  try {
    const response = await apiClient.get<IvfBiopsy[]>('/ivf-lab/biopsies', { params });
    return response.data || [];
  } catch {
    console.warn('Failed to fetch biopsies');
    return [];
  }
};

export const saveBiopsy = async (data: Partial<IvfBiopsy>) => {
  const response = await apiClient.post<IvfBiopsy>('/ivf-lab/biopsies', data);
  return response.data;
};

export const getIvfDashboard = async (): Promise<IvfDashboard> => {
  try {
    const response = await apiClient.get<IvfDashboard>('/ivf-lab/dashboard');
    return response.data;
  } catch {
    console.warn('Failed to fetch IVF dashboard');
    return { activeCycles: 0, frozenEmbryos: 0, spermSamples: 0, transfersThisMonth: 0, successRate: 0, totalCouples: 0, completedCycles: 0 };
  }
};

export const getDailyReport = async (date?: string): Promise<IvfDailyReport> => {
  try {
    const response = await apiClient.get<IvfDailyReport>('/ivf-lab/daily-report', { params: { date } });
    return response.data;
  } catch {
    console.warn('Failed to fetch IVF daily report');
    return { date, items: [] };
  }
};
